### PR TITLE
Add generated informers

### PR DESCRIPTION
**Pull request by ncdc**
Add informer-gen and the informers it generates. We'll do follow-up PRs to convert everything currently using the hand-written informers to the generated ones.

TODO:

- [x] switch to `GroupVersionResource`
- [x] finish godoc

@deads2k @caesarxuchao @sttts @liggitt

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37643)

File: cmd/libs/go2idl/informer-gen/generators/customargs.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package generators
+
+import "github.com/spf13/pflag"
+
+type CustomArgs struct {
+	VersionedClientSetPackage string
```
Comment (ncdc) [inline cmd/libs/go2idl/informer-gen/generators/customargs.go]: What two?

File: pkg/client/informers/informers_generated/generic.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+// ForResource gives generic access to a shared informer of the matching type
+// TODO extend this to unknown resources with a client pool
+func (f *sharedInformerFactory) ForResource(resource schema.GroupResource) (GenericInformer, error) {
+	switch resource {
+	// Group=Apps, Version=InternalVersion
+	case apps.Resource("statefulsets"):
+		return &genericInformer{resource: resource, informer: f.Apps().InternalVersion().StatefulSets().Informer()}, nil
+
+		// Group=Apps, Version=V1beta1
```
Comment (deads2k) [inline pkg/client/informers/informers_generated/generic.go]: comment is wrong: `apps.k8s.io` right?

Comment (ncdc) [inline pkg/client/informers/informers_generated/generic.go]: I don't know if I have any way to get that value

File: pkg/client/informers/informers_generated/generic.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	return f.informer
+}
+
+func (f *genericInformer) Lister() cache.GenericLister {
+	return cache.NewGenericLister(f.Informer().GetIndexer(), f.resource)
+}
+
+// ForResource gives generic access to a shared informer of the matching type
+// TODO extend this to unknown resources with a client pool
+func (f *sharedInformerFactory) ForResource(resource schema.GroupResource) (GenericInformer, error) {
```
Comment (deads2k) [inline pkg/client/informers/informers_generated/generic.go]: If you have a single factory with both internal and external types, you need a full `GroupVersionResource`

**Comment by deads2k** [comment]
Need to fix up the generic access function.  lgtm otherwise.

**Comment by ncdc** [comment]
Will do

File: cmd/libs/go2idl/informer-gen/generators/factory.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  }
+  informer = newFunc(f.versionedClient, f.defaultResync)
+  f.informers[informerType] = informer
+
+  return informer
+}
+`
+
+var sharedInformerFactoryInterface = `
+// SharedInformerFactory TODO
```
Comment (caesarxuchao) [inline cmd/libs/go2idl/informer-gen/generators/factory.go]: Finish the comment?

**Comment by deads2k** [comment]
lgtm

File: pkg/client/informers/informers_generated/generic.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/client/informers/informers_generated/generic.go]: How about using the GroupName variable? https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/register.go#L26

**Comment by caesarxuchao** [comment]
I don't have further comments. Thanks.

File: pkg/client/informers/informers_generated/generic.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/client/informers/informers_generated/generic.go]: You are right. I double checked the client-gen code, it uses the full group name [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/release_1_5/typed/authentication/v1beta1/authentication_client.go#L71), and it's parsed from a tag [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/client-gen/generators/generator_for_group.go#L83-L85), the tag is defined [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/v1beta1/doc.go#L19). The mechanism looks quite hacky though.

**Comment by ncdc** [comment]
Squashed

**Comment by deads2k** [comment]
> Squashed

Looking forward to this.

**Comment by ncdc** [comment]
@k8s-bot gce etcd3 e2e test this issue: #26955

**Comment by ncdc** [comment]
@k8s-bot kops aws e2e test this

**Pull request merged**